manager/src/main/java/org/openremote/manager/asset/ForecastService.java

```diff
@@ -0,0 +1,858 @@
+                long pastPeriod = config.getPastPeriod().toMillis();
+                long forecastPeriod = config.getForecastPeriod().toMillis();
+                long totalForecastPeriod = Math.min(forecastPeriod * config.getForecastCount(), pastPeriod);
+
+                for (int i = config.getPastCount(); i >= 1; i--) {
+                    long timestampMin = now - (pastPeriod * i) + offset;
+                    long timestampMax = now - (pastPeriod * i) + totalForecastPeriod + offset;
+
+                    datapointBuckets.add(new DatapointBucket(timestampMin, timestampMax));
+
+                    query.setParameter("timestampMin" + i, new Date(timestampMin));
+                    query.setParameter("timestampMax" + i, new Date(timestampMax));
+                }
+                return query.getResultList();
+            });
+
+            datapoints.forEach(datapoint -> {
+                datapointBuckets.stream()
+                    .filter(bucket -> bucket.isInTimeRange(datapoint.getTimestamp()))
+                    .findFirst()
+                    .ifPresent(bucket -> bucket.add(datapoint));
+            });
+
+            return datapointBuckets;
+        }
+
+        private List<Long> loadForecastTimestampsFromDb(AttributeRef attributeRef, long now) {
+            List<AssetPredictedDatapoint> datapoints = assetPredictedDatapointService.getDatapoints(attributeRef);
+            List<Long> timestamps = datapoints
+                .stream()
+                .map(datapoint -> datapoint.getTimestamp())
+                .filter(timestamp -> timestamp >= now)
+                .sorted()
+                .collect(Collectors.toList());;
+            return timestamps;
+        }
+    }
+
+    public static class ForecastAttribute {
+
+        private String assetId;
+        private AttributeRef attributeRef;
+        private Attribute<?> attribute;
+        private ForecastConfiguration config;
+        private List<Long> forecastTimestamps = new ArrayList<>();
+
+        public ForecastAttribute(Asset<?> asset, Attribute<?> attribute) {
+            this(asset.getId(), attribute);
+        }
+
+        public ForecastAttribute(String assetId, Attribute<?> attribute) {
+            requireNonNullAndNonEmpty(assetId);
+            if (attribute == null) {
+                throw new IllegalArgumentException("Attribute cannot be null");
+            }
+            this.assetId = assetId;
+            this.attribute = attribute;
+            this.attributeRef = new AttributeRef(assetId, attribute.getName());
+            this.config = attribute.getMetaValue(FORECAST).orElse(null);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            ForecastAttribute that = (ForecastAttribute) o;
+            return assetId.equals(that.assetId) && attribute.getName().equals(that.attribute.getName());
```

Comment:
`Objects` class has great helpers to make overriding `equals` and `hashcode` very simple...IntellIJ can generate them for you as well